# Dealing with imbalanced dataset

Imbalanced dataset

How to resolve imbalancing

      Upsampling
    
      Downsampling

Introduce KNearest Neighbor

HYper parameter tuning using GridSearchCV

Create the best model

Conclude

## Imbalanced dataset

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
bank=pd.read_csv('bank.csv')

In [4]:
bank

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

In [5]:
bank['y'].value_counts()

no     39922
yes     5289
Name: y, dtype: int64

In [6]:
5289/(5289+39922)

0.11698480458295547

Two classes:

Yes, class. >>> 11.7%

No, class >>>> 88.3%

This is an imabalanced dataset

## Resolving imbalancing

### Upsampling 

### Downsampling

## Splitting the data

In [7]:
bank_no=bank[bank['y']=='no']
bank_no.shape

(39922, 17)

In [8]:
bank_no.head(25)

age           job   marital  education default  balance housing loan  \
0    58    management   married   tertiary      no     2143     yes   no   
1    44    technician    single  secondary      no       29     yes   no   
2    33  entrepreneur   married  secondary      no        2     yes  yes   
3    47   blue-collar   married    unknown      no     1506     yes   no   
4    33       unknown    single    unknown      no        1      no   no   
5    35    management   married   tertiary      no      231     yes   no   
6    28    management    single   tertiary      no      447     yes  yes   
7    42  entrepreneur  divorced   tertiary     yes        2     yes   no   
8    58       retired   married    primary      no      121     yes   no   
9    43    technician    single  secondary      no      593     yes   no   
10   41        admin.  divorced  secondary      no      270     yes   no   
11   29        admin.    single  secondary      no      390     yes   no   
12   53    technician   married  secondary      no        6     yes   no   
13   58    technician   married    unknown      no       71     yes   no   
14   57      services   married  secondary      no      162     yes   no   
15   51       retired   married    primary      no      229     yes   no   
16   45        admin.    single    unknown      no       13     yes   no   
17   57   blue-collar   married    primary      no       52     yes   no   
18   60       retired   married    primary      no       60     yes   no   
19   33      services   married  secondary      no        0     yes   no   
20   28   blue-collar   married  secondary      no      723     yes  yes   
21   56    management   married   tertiary      no      779     yes   no   
22   32   blue-collar    single    primary      no       23     yes  yes   
23   25      services   married  secondary      no       50     yes   no   
24   40       retired   married    primary      no        0     yes  yes   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0   unknown    5   may       261         1     -1         0  unknown  no  
1   unknown    5   may       151         1     -1         0  unknown  no  
2   unknown    5   may        76         1     -1         0  unknown  no  
3   unknown    5   may        92         1     -1         0  unknown  no  
4   unknown    5   may       198         1     -1         0  unknown  no  
5   unknown    5   may       139         1     -1         0  unknown  no  
6   unknown    5   may       217         1     -1         0  unknown  no  
7   unknown    5   may       380         1     -1         0  unknown  no  
8   unknown    5   may        50         1     -1         0  unknown  no  
9   unknown    5   may        55         1     -1         0  unknown  no  
10  unknown    5   may       222         1     -1         0  unknown  no  
11  unknown    5   may       137         1     -1         0  unknown  no  
12  unknown    5   may       517         1     -1         0  unknown  no  
13  unknown    5   may        71         1     -1         0  unknown  no  
14  unknown    5   may       174         1     -1         0  unknown  no  
15  unknown    5   may       353         1     -1         0  unknown  no  
16  unknown    5   may        98         1     -1         0  unknown  no  
17  unknown    5   may        38         1     -1         0  unknown  no  
18  unknown    5   may       219         1     -1         0  unknown  no  
19  unknown    5   may        54         1     -1         0  unknown  no  
20  unknown    5   may       262         1     -1         0  unknown  no  
21  unknown    5   may       164         1     -1         0  unknown  no  
22  unknown    5   may       160         1     -1         0  unknown  no  
23  unknown    5   may       342         1     -1         0  unknown  no  
24  unknown    5   may       181         1     -1         0  unknown  no

In [9]:
bank_yes=bank[bank['y']=='yes']
bank_yes.shape

(5289, 17)

In [10]:
bank_yes.head(25)

age          job   marital  education default  balance housing loan  \
83     59       admin.   married  secondary      no     2343     yes   no   
86     56       admin.   married  secondary      no       45      no   no   
87     41   technician   married  secondary      no     1270     yes   no   
129    55     services   married  secondary      no     2476     yes   no   
168    54       admin.   married   tertiary      no      184      no   no   
270    42   management    single   tertiary      no        0     yes  yes   
388    56   management   married   tertiary      no      830     yes  yes   
390    60      retired  divorced  secondary      no      545     yes   no   
446    39   technician    single    unknown      no    45248     yes   no   
457    37   technician   married  secondary      no        1     yes   no   
474    34       admin.   married  secondary      no      869      no   no   
512    55   unemployed  divorced  secondary      no      387     yes   no   
569    28     services    single  secondary      no     5090     yes   no   
624    30   technician   married  secondary      no      484     yes   no   
694    36   technician   married  secondary      no      368     yes  yes   
849    37       admin.    single  secondary      no      245     yes  yes   
877    45  blue-collar   married  secondary      no      154     yes   no   
891    53     services  divorced    primary      no     -291     yes  yes   
897    38       admin.    single  secondary      no      100     yes   no   
1087   30  blue-collar   married  secondary      no      309     yes   no   
1124   29   management   married   tertiary      no      199     yes  yes   
1140   46  blue-collar    single   tertiary      no      460     yes   no   
1172   31   technician    single   tertiary      no      703     yes   no   
1173   39  blue-collar   married    primary      no       70      no   no   
1193   35   management  divorced   tertiary      no     3837     yes   no   

      contact  day month  duration  campaign  pdays  previous poutcome    y  
83    unknown    5   may      1042         1     -1         0  unknown  yes  
86    unknown    5   may      1467         1     -1         0  unknown  yes  
87    unknown    5   may      1389         1     -1         0  unknown  yes  
129   unknown    5   may       579         1     -1         0  unknown  yes  
168   unknown    5   may       673         2     -1         0  unknown  yes  
270   unknown    5   may       562         2     -1         0  unknown  yes  
388   unknown    6   may      1201         1     -1         0  unknown  yes  
390   unknown    6   may      1030         1     -1         0  unknown  yes  
446   unknown    6   may      1623         1     -1         0  unknown  yes  
457   unknown    6   may       608         1     -1         0  unknown  yes  
474   unknown    6   may      1677         1     -1         0  unknown  yes  
512   unknown    6   may       918         1     -1         0  unknown  yes  
569   unknown    6   may      1297         3     -1         0  unknown  yes  
624   unknown    6   may       703         1     -1         0  unknown  yes  
694   unknown    6   may      1597         2     -1         0  unknown  yes  
849   unknown    7   may       732         2     -1         0  unknown  yes  
877   unknown    7   may      1138         1     -1         0  unknown  yes  
891   unknown    7   may       591         1     -1         0  unknown  yes  
897   unknown    7   may       786         1     -1         0  unknown  yes  
1087  unknown    7   may      1574         2     -1         0  unknown  yes  
1124  unknown    7   may      1689         4     -1         0  unknown  yes  
1140  unknown    7   may      1102         2     -1         0  unknown  yes  
1172  unknown    8   may       943         2     -1         0  unknown  yes  
1173  unknown    8   may       813         1     -1         0  unknown  yes  
1193  unknown    8   may      1084         1     -1  

## Upsampling

In [11]:
from sklearn.utils import resample

bank_yes_up=resample(bank_yes,n_samples=15000,replace=True,random_state=100)

bank_yes_up.head(25)

age           job   marital  education default  balance housing loan  \
42571   38        admin.   married  secondary      no    11303      no   no   
3190    41   blue-collar   married  secondary      no     1384     yes   no   
10049   42  entrepreneur   married   tertiary      no     5345      no   no   
31962   31   blue-collar   married  secondary      no     1406     yes  yes   
43024   51    management   married   tertiary      no      346      no   no   
18015   55    management   married   tertiary      no     -375      no   no   
44802   31    management    single   tertiary      no     3340      no   no   
39724   43        admin.   married  secondary      no      132      no   no   
43431   78       retired  divorced    primary      no     1389      no   no   
31180   25    technician    single  secondary      no     1231     yes   no   
41153   67       retired  divorced   tertiary      no      443      no   no   
32027   36    technician   married  secondary      no    15485      no   no   
26961   47    technician   married  secondary      no        0      no   no   
40152   30    technician    single   tertiary      no        0     yes   no   
30176   69    management   married   tertiary      no      840      no   no   
36173   40      services   married  secondary      no     -192     yes   no   
32028   33    management   married   tertiary      no        0     yes   no   
36692   51   blue-collar   married  secondary      no      518     yes   no   
44547   44    management   married   tertiary      no     1791      no   no   
40862   32    technician   married  secondary      no      484     yes   no   
44041   48        admin.    single  secondary      no     1544     yes   no   
21224   47    management   married   tertiary      no      682      no   no   
38732   32   blue-collar    single  secondary      no      217     yes   no   
42785   54        admin.   married  secondary      no        0      no   no   
39972   34    technician   married   tertiary      no      127      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
42571   cellular   28   dec       473         2    216         2  failure  yes  
3190     unknown   15   may      1162         4     -1         0  unknown  yes  
10049    unknown   11   jun       878         3     -1         0  unknown  yes  
31962   cellular   13   apr      1091         2     -1         0  unknown  yes  
43024   cellular   12   feb       122         1     92         5  success  yes  
18015   cellular   30   jul       814         2     -1         0  unknown  yes  
44802   cellular   15   sep       213         2    469         3  success  yes  
39724   cellular   27   may       187         2     71         1  success  yes  
43431   cellular    8   apr       335         1     -1         0  unknown  yes  
31180   cellular   27   feb       412         5     -1         0  unknown  yes  
41153   cellular   18   aug       441         1     -1         0  unknown  yes  
32027   cellular   14   apr       461         1     -1         0  unknown  yes  
26961   cellular   21   nov       591         1     10         1  failure  yes  
40152   cellular    5   jun       159         2     -1         0  unknown  yes  
30176  telephone    5   feb       128         3     -1         0  unknown  yes  
36173   cellular   11   may       666         1     -1         0  unknown  yes  
32028   cellular   14   apr       535         3    328         1  failure  yes  
36692   cellular   12   may       918         1     -1         0  unknown  yes  
44547  telephone   12   aug       201         1    182         2  success  yes  
40862   cellular   12   aug       668         2    463         1  success  yes  
44041  telephone   30   jun       263         1    450         2  failure  yes  
21224   cellular   18   aug       638         6     -1         0  unknown  yes  
38732   cellular   15   may       692         3     -1         0  unknown  yes  
42785   

In [12]:
bank_yes_up.shape

(15000, 17)

## Downsampling

On majority dataset

It reduces the contribution

In [13]:
bank_no_down=resample(bank_no,replace=False,n_samples=25000,random_state=100)
bank_no_down.shape

(25000, 17)

In [14]:
bank_no_down.head(25)

age           job   marital  education default  balance housing loan  \
9789    39   blue-collar   married   tertiary      no    10483      no   no   
44899   77       retired   married   tertiary      no        0      no   no   
2331    28   blue-collar   married  secondary      no      -95     yes   no   
15271   41    management    single   tertiary      no        0     yes   no   
10642   38      services  divorced  secondary      no      439      no   no   
31346   57   blue-collar   married    primary      no     1610      no   no   
15867   41   blue-collar   married  secondary     yes     -750     yes  yes   
24316   48    technician   married  secondary      no       32     yes  yes   
2401    27   blue-collar    single  secondary      no       71     yes   no   
8299    47        admin.   married    primary      no       68     yes   no   
28154   26    technician    single  secondary      no      152      no  yes   
22390   49    management    single   tertiary      no        0      no   no   
28844   57    management   married   tertiary      no        0      no  yes   
29620   54   blue-collar   married  secondary      no      175     yes   no   
11932   46    management   married   tertiary      no      974     yes   no   
24022   37    technician   married  secondary      no       41      no   no   
8968    41    management   married   tertiary      no        5      no   no   
12208   36       student    single  secondary      no      904      no   no   
13139   28     housemaid   married  secondary      no     1294     yes   no   
32136   47    unemployed  divorced  secondary     yes      342     yes   no   
41998   55    technician   married  secondary      no        0     yes  yes   
32473   41        admin.    single  secondary      no     5219     yes   no   
3878    57  entrepreneur   married   tertiary     yes     -524     yes   no   
11630   39   blue-collar   married    primary      no     1046      no   no   
2644    47    technician   married  secondary      no      449     yes  yes   

         contact  day month  duration  campaign  pdays  previous poutcome   y  
9789     unknown    9   jun       218         1     -1         0  unknown  no  
44899   cellular   27   sep       990         4     -1         0  unknown  no  
2331     unknown   13   may       200         1     -1         0  unknown  no  
15271   cellular   17   jul       509         7     -1         0  unknown  no  
10642    unknown   16   jun        50         1     -1         0  unknown  no  
31346  telephone   16   mar       147         1     -1         0  unknown  no  
15867   cellular   21   jul       251         3     -1         0  unknown  no  
24316   cellular   17   nov       110         2     -1         0  unknown  no  
2401     unknown   13   may       174         1     -1         0  unknown  no  
8299     unknown    2   jun       754         2     -1         0  unknown  no  
28154   cellular   29   jan        90         1     -1         0  unknown  no  
22390   cellular   22   aug       205         4     -1         0  unknown  no  
28844   cellular   30   jan       209         2    197         2    other  no  
29620   cellular    3   feb        94         4    271         4  failure  no  
11932    unknown   20   jun       652         3     -1         0  unknown  no  
24022   cellular   29   aug       127         5     -1         0  unknown  no  
8968     unknown    5   jun       178         1     -1         0  unknown  no  
12208    unknown   20   jun        10         8     -1         0  unknown  no  
13139   cellular    8   jul       401         2     -1         0  unknown  no  
32136   cellular   15   apr       177         1     -1         0  unknown  no  
41998   cellular   27   oct        88         1     -1         0  unknown  no  
32473   cellular   17   apr        46         4     -1         0  unknown  no  
3878     unknown   16   may       199         4     -1         0  unknown  no  
11630    unknown   19   jun     

## Creating a new dataset

In [16]:
bank_new=pd.concat([bank_no_down,bank_yes_up])
bank_new.shape

(40000, 17)

In [17]:
bank_new.head(25)

age           job   marital  education default  balance housing loan  \
9789    39   blue-collar   married   tertiary      no    10483      no   no   
44899   77       retired   married   tertiary      no        0      no   no   
2331    28   blue-collar   married  secondary      no      -95     yes   no   
15271   41    management    single   tertiary      no        0     yes   no   
10642   38      services  divorced  secondary      no      439      no   no   
31346   57   blue-collar   married    primary      no     1610      no   no   
15867   41   blue-collar   married  secondary     yes     -750     yes  yes   
24316   48    technician   married  secondary      no       32     yes  yes   
2401    27   blue-collar    single  secondary      no       71     yes   no   
8299    47        admin.   married    primary      no       68     yes   no   
28154   26    technician    single  secondary      no      152      no  yes   
22390   49    management    single   tertiary      no        0      no   no   
28844   57    management   married   tertiary      no        0      no  yes   
29620   54   blue-collar   married  secondary      no      175     yes   no   
11932   46    management   married   tertiary      no      974     yes   no   
24022   37    technician   married  secondary      no       41      no   no   
8968    41    management   married   tertiary      no        5      no   no   
12208   36       student    single  secondary      no      904      no   no   
13139   28     housemaid   married  secondary      no     1294     yes   no   
32136   47    unemployed  divorced  secondary     yes      342     yes   no   
41998   55    technician   married  secondary      no        0     yes  yes   
32473   41        admin.    single  secondary      no     5219     yes   no   
3878    57  entrepreneur   married   tertiary     yes     -524     yes   no   
11630   39   blue-collar   married    primary      no     1046      no   no   
2644    47    technician   married  secondary      no      449     yes  yes   

         contact  day month  duration  campaign  pdays  previous poutcome   y  
9789     unknown    9   jun       218         1     -1         0  unknown  no  
44899   cellular   27   sep       990         4     -1         0  unknown  no  
2331     unknown   13   may       200         1     -1         0  unknown  no  
15271   cellular   17   jul       509         7     -1         0  unknown  no  
10642    unknown   16   jun        50         1     -1         0  unknown  no  
31346  telephone   16   mar       147         1     -1         0  unknown  no  
15867   cellular   21   jul       251         3     -1         0  unknown  no  
24316   cellular   17   nov       110         2     -1         0  unknown  no  
2401     unknown   13   may       174         1     -1         0  unknown  no  
8299     unknown    2   jun       754         2     -1         0  unknown  no  
28154   cellular   29   jan        90         1     -1         0  unknown  no  
22390   cellular   22   aug       205         4     -1         0  unknown  no  
28844   cellular   30   jan       209         2    197         2    other  no  
29620   cellular    3   feb        94         4    271         4  failure  no  
11932    unknown   20   jun       652         3     -1         0  unknown  no  
24022   cellular   29   aug       127         5     -1         0  unknown  no  
8968     unknown    5   jun       178         1     -1         0  unknown  no  
12208    unknown   20   jun        10         8     -1         0  unknown  no  
13139   cellular    8   jul       401         2     -1         0  unknown  no  
32136   cellular   15   apr       177         1     -1         0  unknown  no  
41998   cellular   27   oct        88         1     -1         0  unknown  no  
32473   cellular   17   apr        46         4     -1         0  unknown  no  
3878     unknown   16   may       199         4     -1         0  unknown  no  
11630    unknown   19   jun     

In [18]:
bank_new.tail(25)

age            job  marital  education default  balance housing loan  \
44474   47     management  married   tertiary      no     2306      no   no   
40456   25         admin.   single  secondary      no      285      no   no   
33943   29        student   single  secondary      no      314      no   no   
42457   49     technician  married  secondary      no      307      no   no   
43416   30     management  married   tertiary      no      102     yes   no   
41849   27     management  married   tertiary      no      893      no   no   
4916    37     technician   single  secondary      no     1211     yes   no   
2882    40         admin.  married  secondary      no     1535     yes   no   
33704   35     management   single   tertiary      no      704      no   no   
39780   26     management   single  secondary      no     2613      no   no   
33803   72        retired  married    primary      no      763      no   no   
7123    35         admin.  married  secondary      no      404     yes   no   
45004   29  self-employed   single   tertiary      no     1579      no   no   
40035   25     management   single   tertiary      no      362      no   no   
28696   41     management   single  secondary      no     3234     yes   no   
45178   34    blue-collar  married  secondary      no      320     yes   no   
43547   29     management   single   tertiary      no     1445      no   no   
31325   44     management  married   tertiary      no      423      no   no   
7851    46     management   single   tertiary      no     2904     yes   no   
42468   49     management  married   tertiary      no      395      no   no   
38934   24       services   single  secondary      no      414     yes   no   
42798   62   entrepreneur  married  secondary      no     3904      no   no   
31333   58  self-employed  married   tertiary      no     5810      no   no   
42557   33     management  married   tertiary      no     1808      no   no   
35533   42     unemployed  married  secondary      no      356     yes   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
44474  telephone    6   aug       239         4     -1         0  unknown  yes  
40456   cellular    3   jul       658         3     -1         0  unknown  yes  
33943   cellular   30   apr       465         1     -1         0  unknown  yes  
42457   cellular    9   dec       854         1     -1         0  unknown  yes  
43416   cellular    5   apr       470         7    426         3    other  yes  
41849   cellular   16   oct       154         2     -1         0  unknown  yes  
4916     unknown   21   may       795         1     -1         0  unknown  yes  
2882     unknown   14   may       704         1     -1         0  unknown  yes  
33704   cellular   21   apr       164         1     -1         0  unknown  yes  
39780   cellular    1   jun       326         1     -1         0  unknown  yes  
33803   cellular   24   apr       384         2     -1         0  unknown  yes  
7123     unknown   29   may       539         1     -1         0  unknown  yes  
45004   cellular   13   oct       190         1     92         5  success  yes  
40035   cellular    3   jun       617         2     -1         0  unknown  yes  
28696   cellular   30   jan      2770         1     -1         0  unknown  yes  
45178   cellular   12   nov       413         1     92         3  success  yes  
43547   cellular   26   apr       328         2    143         2  success  yes  
31325   cellular   12   mar      2330         5     -1         0  unknown  yes  
7851     unknown   30   may       984         2     -1         0  unknown  yes  
42468   cellular   10   dec       336         5     -1         0  unknown  yes  
38934   cellular   18   may       493         1    370         1    other  yes  
42798  telephone   29   jan       403         2     -1         0  unknown  yes  
31333   cellular   12   mar       139         1     -1         0  unknown  yes  
42557   

## Shuffling the data


In [19]:
from sklearn.utils import shuffle

bank_new=shuffle(bank_new)
bank_new.head(25)

age           job  marital  education default  balance housing loan  \
23153   38    technician  married   tertiary      no     -363      no   no   
14938   42      services  married  secondary      no     3465     yes  yes   
4986    33    technician   single  secondary      no      296     yes   no   
22002   30    technician   single  secondary      no      790      no   no   
15922   35    technician  married  secondary      no      561     yes   no   
33793   51    management  married   tertiary      no     4917      no   no   
23248   53    unemployed  married    primary      no       49      no   no   
15538   34    technician  married  secondary      no     6324     yes   no   
44808   36    technician  married   tertiary      no      376      no   no   
1520    37    unemployed   single  secondary      no      381     yes   no   
41554   73       retired  married    unknown      no      739      no   no   
33725   63        admin.  married  secondary      no      -40      no   no   
23533   55     housemaid  married    primary      no     1233      no   no   
22439   30    management   single   tertiary      no      370      no   no   
26507   44    management   single   tertiary      no     3812      no   no   
39763   25    technician   single  secondary      no     1199      no   no   
40039   32   blue-collar  married  secondary      no      889     yes   no   
32314   36        admin.   single  secondary      no     1414     yes   no   
26544   36    management   single   tertiary      no     2032      no  yes   
36919   58  entrepreneur  married    primary      no     1153     yes  yes   
43739   33    management   single   tertiary      no     2364      no   no   
13515   44  entrepreneur  married   tertiary      no     1808     yes   no   
39925   28    technician   single   tertiary      no     4987      no   no   
21596   36    technician  married  secondary      no      632     yes   no   
25925   32    technician   single   tertiary      no      550      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
23153   cellular   26   aug       101         2     -1         0  unknown   no  
14938   cellular   16   jul      1039         2     -1         0  unknown  yes  
4986     unknown   21   may       226         1     -1         0  unknown   no  
22002   cellular   20   aug       696         3     -1         0  unknown   no  
15922   cellular   22   jul        95         1     -1         0  unknown   no  
33793   cellular   23   apr       113         1     -1         0  unknown  yes  
23248   cellular   27   aug       254         8     -1         0  unknown   no  
15538   cellular   21   jul       190         4     -1         0  unknown   no  
44808   cellular   16   sep       173         1     -1         0  unknown  yes  
1520     unknown    8   may       985         2     -1         0  unknown  yes  
41554   cellular   11   sep       210         1     -1         0  unknown  yes  
33725   cellular   22   apr        99         3     -1         0  unknown   no  
23533   cellular   28   aug        97         2     -1         0  unknown   no  
22439   cellular   22   aug        64         2     -1         0  unknown   no  
26507   cellular   20   nov       172         2     -1         0  unknown   no  
39763   cellular    1   jun       287         1     -1         0  unknown  yes  
40039   cellular    3   jun      1076         2     85         3    other  yes  
32314   cellular   16   apr       244         1     -1         0  unknown   no  
26544   cellular   20   nov       736         1     -1         0  unknown  yes  
36919  telephone   12   may       206         6    337         6  failure   no  
43739   cellular   17   may       508         1     -1         0  unknown  yes  
13515   cellular    9   jul       127         1     -1         0  unknown   no  
39925   cellular    2   jun       924         2    113        21  success  yes  
21596   cellular   19   aug       

## Separating to target and features

In [20]:
# Target
y=bank_new['y']
y

23153     no
14938    yes
4986      no
22002     no
15922     no
        ... 
42941    yes
10094     no
5476      no
43460    yes
15957     no
Name: y, Length: 40000, dtype: object

In [21]:
# Features

X=bank_new.drop(['y'],axis=1)
X

age         job   marital  education default  balance housing loan  \
23153   38  technician   married   tertiary      no     -363      no   no   
14938   42    services   married  secondary      no     3465     yes  yes   
4986    33  technician    single  secondary      no      296     yes   no   
22002   30  technician    single  secondary      no      790      no   no   
15922   35  technician   married  secondary      no      561     yes   no   
...    ...         ...       ...        ...     ...      ...     ...  ...   
42941   45      admin.   married  secondary      no     1206      no   no   
10094   53     retired  divorced  secondary      no      -24      no   no   
5476    39  technician  divorced  secondary      no     1222     yes   no   
43460   30  management    single   tertiary      no     3779      no   no   
15957   33  technician    single   tertiary      no     1257     yes   no   

         contact  day month  duration  campaign  pdays  previous poutcome  
23153   cellular   26   aug       101         2     -1         0  unknown  
14938   cellular   16   jul      1039         2     -1         0  unknown  
4986     unknown   21   may       226         1     -1         0  unknown  
22002   cellular   20   aug       696         3     -1         0  unknown  
15922   cellular   22   jul        95         1     -1         0  unknown  
...          ...  ...   ...       ...       ...    ...       ...      ...  
42941  telephone    8   feb       309         1    446         3  failure  
10094    unknown   11   jun       303         1     -1         0  unknown  
5476     unknown   23   may       451         4     -1         0  unknown  
43460   cellular   13   apr      1093         1     -1         0  unknown  
15957   cellular   22   jul        66         1     -1         0  unknown  

[40000 rows x 16 columns]

## Converting categorical features to numeric

In [22]:
X_new=pd.get_dummies(X)
X_new

age  balance  day  duration  campaign  pdays  previous  job_admin.  \
23153   38     -363   26       101         2     -1         0           0   
14938   42     3465   16      1039         2     -1         0           0   
4986    33      296   21       226         1     -1         0           0   
22002   30      790   20       696         3     -1         0           0   
15922   35      561   22        95         1     -1         0           0   
...    ...      ...  ...       ...       ...    ...       ...         ...   
42941   45     1206    8       309         1    446         3           1   
10094   53      -24   11       303         1     -1         0           0   
5476    39     1222   23       451         4     -1         0           0   
43460   30     3779   13      1093         1     -1         0           0   
15957   33     1257   22        66         1     -1         0           0   

       job_blue-collar  job_entrepreneur  ...  month_jun  month_mar  \
23153                0                 0  ...          0          0   
14938                0                 0  ...          0          0   
4986                 0                 0  ...          0          0   
22002                0                 0  ...          0          0   
15922                0                 0  ...          0          0   
...                ...               ...  ...        ...        ...   
42941                0                 0  ...          0          0   
10094                0                 0  ...          1          0   
5476                 0                 0  ...          0          0   
43460                0                 0  ...          0          0   
15957                0                 0  ...          0          0   

       month_may  month_nov  month_oct  month_sep  poutcome_failure  \
23153          0          0          0          0                 0   
14938          0          0          0          0                 0   
4986           1          0          0          0                 0   
22002          0          0          0          0                 0   
15922          0          0          0          0                 0   
...          ...        ...        ...        ...               ...   
42941          0          0          0          0                 1   
10094          0          0          0          0                 0   
5476           1          0          0          0                 0   
43460          0          0          0          0                 0   
15957          0          0          0          0                 0   

       poutcome_other  poutcome_success  poutcome_unknown  
23153               0                 0                 1  
14938               0                 0                 1  
4986                0                 0                 1  
22002               0                 0                 1  
15922               0                 0                 1  
...               ...               ...               ...  
42941               0                 0                 0  
10094               0                 0                 1  
5476                0                 0                 1  
43460               0                 0                 1  
15957               0                 0                 1  

[40000 rows x 51 columns]

## Standardisation of features

In [24]:
from sklearn.preprocessing import StandardScaler

scaler=StandardScaler()

X_scaled=scaler.fit_transform(X_new)
X_scaled

array([[-0.27256415, -0.55825794,  1.23230374, ..., -0.21921049,
        -0.28888227,  0.55166443],
       [ 0.07458499,  0.58157581,  0.04131826, ..., -0.21921049,
        -0.28888227,  0.55166443],
       [-0.70650057, -0.36203265,  0.636811  , ..., -0.21921049,
        -0.28888227,  0.55166443],
       ...,
       [-0.18577686, -0.08630484,  0.87500809, ..., -0.21921049,
        -0.28888227,  0.55166443],
       [-0.96686243,  0.67507314, -0.31597738, ..., -0.21921049,
        -0.28888227,  0.55166443],
       [-0.70650057, -0.07588316,  0.75590955, ..., -0.21921049,
        -0.28888227,  0.55166443]])

## Splitting to train and test

In [25]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X_scaled,y,test_size=0.2,random_state=100)

X_train.shape,X_test.shape,y_train.shape,y_test.shape

((32000, 51), (8000, 51), (32000,), (8000,))

## Model building - K Nearest Neighbor

In [26]:
from sklearn.neighbors import KNeighborsClassifier

knn=KNeighborsClassifier()

knn.fit(X_train,y_train)

KNeighborsClassifier()

## Performance analysis

In [28]:
from sklearn.metrics import confusion_matrix, classification_report

report=classification_report(y_test, knn.predict(X_test))
cm=confusion_matrix(y_test, knn.predict(X_test))

print(' The CM:\n', cm)
print(' The report:\n', report)

 The CM:
 [[4340  677]
 [ 565 2418]]
 The report:
               precision    recall  f1-score   support

          no       0.88      0.87      0.87      5017
         yes       0.78      0.81      0.80      2983

    accuracy                           0.84      8000
   macro avg       0.83      0.84      0.84      8000
weighted avg       0.85      0.84      0.85      8000



## Hyper parameter tuning using GridSearchCV

In [30]:
from sklearn.model_selection import GridSearchCV

knn_gs=GridSearchCV(knn, {'n_neighbors': range(2,8)})

In [32]:
knn_gs.fit(X_train,y_train)

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': range(2, 8)})

In [33]:
knn_gs.best_params_

{'n_neighbors': 2}

## Building the best model


In [35]:
knn_best=KNeighborsClassifier(n_neighbors=2)
knn_best.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=2)

In [36]:
report=classification_report(y_test,knn_best.predict(X_test))

print(' Report:\n', report)

 Report:
               precision    recall  f1-score   support

          no       0.89      0.92      0.90      5017
         yes       0.86      0.81      0.83      2983

    accuracy                           0.88      8000
   macro avg       0.87      0.86      0.87      8000
weighted avg       0.88      0.88      0.88      8000

